[MTKA-1389] Export ACM Blueprints via WP-CLI

nickcernis opened the pull request:
Adds a `wp acm blueprint export` WP-CLI command to dump the current state of the site as a zipped ACM blueprint.

By default this saves:
- ACM models and fields.
- ACM taxonomies.
- ACM entries.
- WP core posts and pages.
- Taxonomy terms.
- Post tags.
- Post meta.
- Media linked to a post via an ACM media field or stored in `_thumbnail_id` post meta.
- ACM relationships.

https://wpengine.atlassian.net/browse/MTKA-1389

## Testing

Includes integration tests for export functionality.

### To test manually

1. Open a shell ready for WP-CLI. (In Local choose Open Site Shell from the site options menu.)
2. Run `wp acm blueprint export --open`.
3. You should see a directory containing your blueprint zip file. It will contain media if exported posts have post meta for a featured image or ACM media field.

To see all options use `wp help acm blueprint export`.

If you wish you can also try:

1. Taking the zip generated by the export.
2. Uploading it somewhere publicly accessible (such as to a WPE WP test site).
3. Removing existing ACM posts/taxonomies from your site. (`wp option delete atlas_content_modeler_post_types && wp option delete atlas_content_modeler_taxonomies`)
4. Restoring models/taxonomies/other state with `wp acm blueprint import https://example.com/path/to/your/blueprint.zip`.

## Screenshots

Defaults:

```sh
> wp acm blueprint export --open
Collecting ACM data.
Collecting ACM models.
Collecting ACM taxonomies.
Collecting posts.
Collecting terms.
Collecting post meta.
Collecting media.
Collecting ACM relationships
Writing acm.json manifest.
Generating zip.
Opening blueprint temp folder.
Success: Blueprint saved to /var/folders/mc/yf37tb4j5bjbwhv6z0m1bcfm0000gp/T/acm/acm-blueprint/acm-blueprint.zip.
```

With custom options:

```sh
> wp acm blueprint export --name="Custom Name" --description="My blueprint is great." --open
Collecting ACM data.
Collecting ACM models.
Collecting ACM taxonomies.
Collecting posts.
Collecting terms.
Collecting post meta.
Collecting media.
Collecting ACM relationships
Writing acm.json manifest.
Generating zip.
Opening blueprint temp folder.
Success: Blueprint saved to /var/folders/mc/yf37tb4j5bjbwhv6z0m1bcfm0000gp/T/acm/custom-name/custom-name.zip.
```

```sh
> wp help acm blueprint export

NAME

  wp acm blueprint export

DESCRIPTION

  Exports an ACM blueprint using the current state of the site.

SYNOPSIS

  wp acm blueprint export [--name] [--description] [--min-wp] [--min-acm] [--version] [--cpt-include] [--open]

  [--name]
    Optional blueprint name. Used in the manifest and zip file name.
    Defaults to “ACM Blueprint” resulting in acm-blueprint.zip.

  [--description]
    Optional description of the blueprint.

  [--min-wp]
    Minimum WordPress version. Defaults to current WordPress version.

  [--min-acm]
    Minimum Atlas Content Modeler plugin version. Defaults to current
    ACM version.

  [--version]
    Optional blueprint version. Defaults to 1.0.

  [--cpt-include]
    Post types to include, separated by commas. Defaults to post, page and
    all registered ACM post types.

  [--open]
    Open the folder containing the generated zip on success (macOS only,
    requires that `shell_exec()` has not been disabled).
```

## Documentation Changes

I plan to file a separate PR or add docs to Google Docs for internal use by other teams.

## Dependant PRs

n/a

nickcernis committed:
chore: generate ACM manifest meta

nickcernis committed:
chore: add type info

nickcernis committed:
chore: write manifest file to temp dir

nickcernis committed:
Merge branch 'feature/wp-cli-blueprint' into mtka-1389/blueprint-export

nickcernis committed:
fix: correct warning logic in import command

nickcernis committed:
fix: adds error handling for manifest write

nickcernis committed:
fix: create manifest write path if not present

nickcernis committed:
chore: collect ACM models and taxonomies

nickcernis committed:
chore: collect posts, pages and ACM entries

nickcernis committed:
fix: pretty print acm.json manifest

nickcernis committed:
chore: collect terms for ACM taxonomies

This currently excludes terms for the
built-in 'post' type (tags/categories).

nickcernis committed:
chore: collect tags for ACM entries

nickcernis committed:
fix: use 'tag' for post tags instead of 'term'

nickcernis committed:
refactor: move variable declaration outside loop

nickcernis committed:
refactor: remove arrow functions

Props @mindctrl, who pointed out these
aren't supported in our min PHP (7.2).

nickcernis committed:
chore: collect post_meta for all entries

nickcernis committed:
chore: collect media for all post meta

nickcernis committed:
chore: collect ACM relationships

nickcernis committed:
chore: write blueprint zip file

nickcernis committed:
fix: correct import process and test zip

The `acm blueprint export` creates zip that decompress into a folder with
the same name as the zip file.

This commit adjusts the import process to:

– Assume the zip file expands to a subfolder of the same name.
– Use a test zip that does not expand into the current directory
   (it expands into a subfolder instead).

nickcernis committed:
fix: correct media export handling

- Only zip media if there is media to zip.
- Include media subfolders in exported paths.

nickcernis committed:
test: verify generate_meta

nickcernis committed:
test: verify collect_posts

nickcernis committed:
test: verify collect_terms

nickcernis committed:
test: verify collect_post_tags

nickcernis committed:
fix: remove other unneeded internal meta

nickcernis committed:
test: verify collect_post_meta

nickcernis committed:
test: verify collect_media

nickcernis committed:
test: verify collect_relationships

nickcernis committed:
test: verify get_acm_temp_dir

nickcernis committed:
test: verify write_manifest

nickcernis committed:
test: verify test_zip_blueprint

nickcernis committed:
refactor: remove unused import

nickcernis committed:
fix: correct broken path in test_zip_blueprint

nickcernis committed:
chore: add changelog message

nickcernis committed:
Merge branch 'feature/wp-cli-blueprint' into mtka-1389/blueprint-export

nickcernis committed:
fix: changelog spacing after merge

nickcernis committed:
refactor: other log errors weren't translatable

matthewguywright commented on includes/wp-cli/class-blueprint.php:
```diff
@@ -188,14 +202,124 @@ public function import( $args, $assoc_args ) {
+	 *
+	 * [--min-acm]
+	 * : Minimum Atlas Content Modeler plugin version. Defaults to current
+	 * ACM version.
```
Nice documentation!

mindctrl commented on includes/wp-cli/class-blueprint.php:
```diff
@@ -188,14 +202,124 @@ public function import( $args, $assoc_args ) {
+	 * [--version]
+	 * : Optional blueprint version. Defaults to 1.0.
+	 *
+	 * [--cpt-include]
```
I wonder if `cpt-include` should be renamed to more accurately reflect its behavior? My thinking is "cpt" being an abbreviation for "custom post type", the name doesn't seem accurate because it's possible to pass any post types to it, not just custom ones, and it defaults to including built-in `post` and `page`. My first thought was `--post-types`. Thoughts?

nickcernis replied:
@mindctrl Nice idea! Amended in https://github.com/wpengine/atlas-content-modeler/pull/425/commits/3f44d885dd8c1ef5af6ad61eea927c4351f2f34d.

rfmeier reviewed (approved):
Awesome work on this!

mindctrl reviewed (commented):
Left a thought about maybe renaming the `cpt-include` option. Otherwise really nice work.

nickcernis committed:
chore: use 'post-types' instead of 'cpt-include'

Core post and page types aren't 'custom' post types.

Props @mindctrl for the suggestion.

nickcernis committed:
fix: clean up previous exports on start

So that previous media folders are removed and not copied to the zip for
subsequent runs that don't collect media.

nickcernis committed:
fix: improve bad blueprint name handling

To catch blueprint names that would result in a blank
folder or zip file name, such as:

```
wp acm blueprint export --open --name='!!!'
```
